packages/itwinui-react/src/utils/hooks/useControlledState.ts

```diff
@@ -4,6 +4,25 @@
  *--------------------------------------------------------------------------------------------*/
 import * as React from 'react';
 
+/**
+ * Get the controlled state if it is defined, otherwise return the uncontrolled state.
+ *
+ * @example
+ * const value = getControlledState(props.value, props.defaultValue);
+ *
+ * @example
+ * const onClickHandler = (clickedValue) => {
+ *   const newState = getControlledState(props.value, clickedValue);
+ *   setState(newState);
+ * };
+ */
+export const getControlledState = <T>(
+  controlledState: T,
+  uncontrolledState: T,
+) => {
+  return controlledState !== undefined ? controlledState : uncontrolledState;
```

Comment:
Why the abstraction?

This is straightforward and immediately super clear:

```js
const newState = props.value !== undefined ? props.value : clickedValue;
```

while this is the same amount of code but now with added indirection:

```js
import { getControlledState } from "../../utils.js";

const newState = getControlledState(props.value, clickedValue)
```

Reply:
I felt that if we change the logic like:

```diff
- const newState = controlledState !== undefined ? controlledState : uncontrolledState;
+ const newState = controlledState != undefined ? controlledState : uncontrolledState;
```

we can do it in just one place. Additionally, I felt that writing it inline might be error prone. E.g. if we accidentally do `!= undefined`/`!= null` instead of `!== undefined`.

---

Maybe this function can have a better name? i.e. Something that signifies "returns the first non-undefined arg from ...args"?

Reply:
The real problem is that you're sidestepping the `useControlledState` hook that does the heavylifting.

I'm trying to review your recent changes from https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1621543479 and I'm not really sure why you brought back the reducer. I was just asking for the `value` vs `index` parts to be changed.

Reply:
> I'm trying to review your recent changes from [#2048 (comment)](https://github.com/iTwin/iTwinUI/pull/2048#discussion_r1621543479) and I'm not really sure why you brought back the reducer. I was just asking for the `value` vs `index` parts to be changed.

Ohh, I see. When I started working on the PR, I think I faced issues with using `useControlledState` with indexes. That is why I had gone with using values instead of indexes. So, when I was working on moving from values back to indexes, I thought to again use the reducer instead of `useControlledState`.

To confirm, since I have now fixed the bug while using the reducer, is that alright? Or I can try going back to using `useControlledState`, but this time with indexes. If so, then I could do this in this PR or maybe another one to avoid delay to the bug fix.

> The real problem is that you're sidestepping the `useControlledState` hook that does the heavylifting.

So, are you suggesting keeping it inline? I'm not sure that's the best approach as I feel it's error prone and can lead to regressions in the future if our tests don't catch the accidental changes in the future in the inline if-statement.

Changed it to inline for now (https://github.com/iTwin/iTwinUI/pull/2048/commits/d8ae400d20c899e147b3e158657d6bc6bfdd1c32) until we come with a better way (if we do so).

Reply:
> To confirm, since I have now fixed the bug while using the reducer, is that alright? Or I can try going back to using `useControlledState`, but this time with indexes. If so, then I could do this in this PR or maybe another one to avoid delay to the bug fix.

I think the code has become harder to follow as a result of extending the reducer. I'm not sure if it would be simpler with `useControlledState` but it's worth trying, just to evaluate. Let me know if this takes too long or if you need help with this.